src/lib/mina_net2/tests/tests.ml

```diff
@@ -0,0 +1,297 @@
+        shutdown ()
+      in
+      Async.Thread_safe.block_on_async_exn (fun () -> test_def)
+
+    let%test_unit "b_stream_c" =
+      let () = Core.Backtrace.elide := false in
+      ignore testmsg ;
+      let test_def =
+        let open Deferred.Let_syntax in
+        let%bind b, c, shutdown = setup_two_nodes "test_stream" in
+        let%bind b_peerid = me b >>| Keypair.to_peer_id in
+        let handler_finished = ref false in
+        let%bind _echo_handler =
+          handle_protocol b ~on_handler_error:`Raise ~protocol:"read_bytes"
+            (fun stream ->
+              let r, w = Stream.pipes stream in
+              let rec go i =
+                if i = 0 then return ()
+                else
+                  let%bind _s =
+                    match%map Pipe.read' ~max_queue_length:1 r with
+                    | `Eof ->
+                        failwith "Eof"
+                    | `Ok q ->
+                        Base.Queue.peek_exn q
+                  in
+                  go (i - 1)
+              in
+              let%map () = go 1000 in
+              Pipe.write_without_pushback w "done" ;
+              Pipe.close w ;
+              handler_finished := true )
+          |> Deferred.Or_error.ok_exn
+        in
+        let%bind stream =
+          open_stream c ~protocol:"read_bytes" b_peerid >>| Or_error.ok_exn
+        in
+        let r, w = Stream.pipes stream in
+        for i = 0 to 999 do
+          Pipe.write_without_pushback w (Printf.sprintf "%d" i)
+        done ;
+        Pipe.close w ;
+        (* HACK: let our messages send before we reset.
+           It would be more principled to add flushing to
+           the stream interface. *)
+        let%bind () = after (Time.Span.of_sec 5.) in
+        let%bind _ = Stream.reset stream in
+        let%bind _msgs = Pipe.read_all r in
+        assert !handler_finished ;
+        shutdown ()
+      in
+      Async.Thread_safe.block_on_async_exn (fun () -> test_def)
+  *)
+
+    let%test_unit "stream" =
+      let () = Core.Backtrace.elide := false in
+      let test_def =
+        let open Deferred.Let_syntax in
+        let%bind b, c, shutdown = setup_two_nodes "test_stream" in
+        let%bind b_peerid = me b >>| Keypair.to_peer_id in
+        let handler_finished = ref false in
+        let%bind () =
+          open_protocol b ~on_handler_error:`Raise ~protocol:"echo"
+            (fun stream ->
+              let r, w = Libp2p_stream.pipes stream in
+              let%map () = Pipe.transfer r w ~f:Fn.id in
+              Pipe.close w ;
+              handler_finished := true)
+          |> Deferred.Or_error.ok_exn
+        in
+        let%bind stream =
+          open_stream c ~protocol:"echo" b_peerid >>| Or_error.ok_exn
+        in
+        let r, w = Libp2p_stream.pipes stream in
+        Pipe.write_without_pushback w testmsg ;
+        Pipe.close w ;
+        (* HACK: let our messages send before we reset.
```

Comment:
couldn't we read the reply from the peer to ensure synchronization instead of 1s wait?

Reply:
I attempted to do this, but the test hangs forever when I do. It's almost like the write doesn't flush until the reset happens, which is odd.